convex/lp/simplex.go

```diff
@@ -0,0 +1,608 @@
+	//
+	// Algorithm:
+	// 1) Compute the "reduced costs" for the non-basic variables. The reduced
+	// costs are the lagrange multipliers of the constraints.
+	// 	 r = cn - an^T * ab^-T * cb
+	// 2) If all of the reduced costs are positive, no improvement is possible,
+	// and the solution is optimal (xn can only increase because of
+	// non-negativity constraints). Otherwise, the solution can be improved and
+	// one element will be exchanged in the basic set.
+	// 3) Choose the x_n with the most negative value of r. Call this value xe.
+	// This variable will be swapped into the basic set.
+	// 4) Increase xe until the next constraint boundary is met. This will happen
+	// when the first element in xb becomes 0. The distance xe can increase before
+	// a given element in xb becomes negative can be found from
+	//	xb = Ab^-1 b - Ab^-1 An xn
+	//     = Ab^-1 b - Ab^-1 Ae xe
+	//     = bhat + d x_e
+	//  xe = bhat_i / - d_i
+	// where Ae is the column of A corresponding to xe.
+	// The constraining basic index is the first index for which this is true,
+	// so remove the element which is min_i (bhat_i / -d_i), assuming d_i is negative.
+	// If no d_i is less than 0, then the problem is unbounded.
+	// 5) If the new xe is 0 (that is, bhat_i == 0), then this location is at
+	// the intersection of several constraints. Use the Bland rule instead
+	// of the rule in step 4 to avoid cycling.
+	for {
+		// Compute reduced costs -- r = cn - an^T ab^-T cb
+		var tmp mat64.Vector
+		err = tmp.SolveVec(ab.T(), cbVec)
+		if err != nil {
+			break
+		}
+		data := make([]float64, n-m)
+		tmp2 := mat64.NewVector(n-m, data)
+		tmp2.MulVec(an.T(), &tmp)
+		floats.SubTo(r, cn, data)
+
+		// Replace the most negative element in the simplex. If there are no
+		// negative entries then the optimal solution has been found.
+		minIdx := floats.MinIdx(r)
+		if r[minIdx] >= -tol {
+			break
+		}
+
+		for i, v := range r {
+			if math.Abs(v) < rRoundTol {
+				r[i] = 0
+			}
+		}
+
+		// Compute the moving distance.
+		err = computeMove(move, minIdx, A, ab, xb, nonBasicIdx)
+		if err != nil {
+			if err == ErrUnbounded {
+				return math.Inf(-1), nil, nil, ErrUnbounded
+			}
+			break
+		}
+
+		// Replace the basic index along the tightest constraint.
+		replace := floats.MinIdx(move)
+		if move[replace] <= 0 {
+			replace, minIdx, err = replaceBland(A, ab, xb, basicIdxs, nonBasicIdx, r, move)
+			if err != nil {
+				if err == ErrUnbounded {
+					return math.Inf(-1), nil, nil, ErrUnbounded
+				}
+				break
+			}
+		}
+
+		// Replace the constrained basicIdx with the newIdx.
+		basicIdxs[replace], nonBasicIdx[minIdx] = nonBasicIdx[minIdx], basicIdxs[replace]
+		cb[replace], cn[minIdx] = cn[minIdx], cb[replace]
+		tmpCol1 := mat64.Col(nil, replace, ab)
+		tmpCol2 := mat64.Col(nil, minIdx, an)
+		ab.SetCol(replace, tmpCol2)
+		an.SetCol(minIdx, tmpCol1)
+
+		// Compute the new xb.
+		xbVec := mat64.NewVector(len(xb), xb)
+		err = xbVec.SolveVec(ab, bVec)
```

Comment:
It seems to me that the matrix factorized here and in the next iteration inside computeMove is the same, so one factorization can be avoided. I'm not sure how much work that would be, a future PR is ok.